### PR TITLE
GEOMESA-2921 Implement a graduated spatial-temporal query guard

### PR DESCRIPTION
**Pull request by jnh5y**
Signed-off-by: Jim Hughes <jnh5y@ccri.com>

### CODE REVIEW

**Comment by jnh5y** [comment]
TODOs:
1.  Refactor two query guards out
2.  Test in a GeoServer
3.  Add documentation

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/z2/package.scala [hunk 1]
```diff
@@ -16,9 +16,9 @@ package object z2 {
 
   case class Z2IndexValues(sfc: Z2SFC,
                            geometries: FilterValues[Geometry],
-                           bounds: Seq[(Double, Double, Double, Double)])
+                           spatialBounds: Seq[(Double, Double, Double, Double)]) extends SpatialIndexValues
 
   case class XZ2IndexValues(sfc: XZ2SFC,
                             geometries: FilterValues[Geometry],
-                            bounds: Seq[(Double, Double, Double, Double)])
+                            spatialBounds: Seq[(Double, Double, Double, Double)]) extends SpatialIndexValues
```
Comment (elahrvivaz) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/z2/package.scala:23]: not sure if this is in the 'public api', but we could leave the `bounds` the same, mark it deprecated, and add in an `override def spatialBounds = bounds` method to fulfill the new SpatialIndexValues trait

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala [hunk 2]
```diff
@@ -160,14 +164,135 @@ object QueryInterceptor extends LazyLogging {
...
+     * 1. Sizes must not be repeated.
+     * 2. Smaller sizes must allow for longer durations.
+     * @param limits Sequence of limits
+     * @return Sorted list of limits or throws an IllegalArgumentException
+     */
+    def evaluateLimits(limits: Seq[SizeAndDuration]): Seq[SizeAndDuration] = {
+      val candidate: Seq[SizeAndDuration] = limits.sortBy(_.sizeLimit)
+      candidate.sliding(2).foreach { seq =>
+        val first = seq.head
+        val second = seq.tail.head
```
Comment (elahrvivaz) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala:207]: might be slightly nicer to pattern match like `.foreach { case Seq(first, second) =>`
one other gotcha is that if the list only has one element, `sliding(2)` will give you a list with one element in it...

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala [hunk 3]
```diff
@@ -160,14 +164,135 @@ object QueryInterceptor extends LazyLogging {
...
+      val spatialExtent = spatialBoundsOption match {
+        case None => 360.0 * 180.0  // Whole world
+        case Some(seq) => seq.map { case  (lx: Double, ly: Double, ux: Double, uy: Double) =>
+          (ux-lx) * (uy-ly)
+        }.sum
+      }
+
+      intervalsOption match {
+        case None => // Q: What does None here mean again?!
+          None       // Returning None means that the query continues.  Is this correct?
```
Comment (elahrvivaz) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala:272]: right

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala [hunk 4]
```diff
@@ -160,14 +164,135 @@ object QueryInterceptor extends LazyLogging {
...
+
+      val spatialExtent = spatialBoundsOption match {
+        case None => 360.0 * 180.0  // Whole world
+        case Some(seq) => seq.map { case  (lx: Double, ly: Double, ux: Double, uy: Double) =>
+          (ux-lx) * (uy-ly)
+        }.sum
+      }
+
+      intervalsOption match {
+        case None => // Q: What does None here mean again?!
```
Comment (elahrvivaz) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala:271]: we might want to re-evaluate the `strategy.values.collect { case v: TemporalIndexValues => v.intervals }`, b/c it makes None ambiguous - it could mean that the index doesn't have a temporal component (the collect doesn't match), or that the temporal component is empty (v.intervals is None). those should probably be handled differently...

Comment (elahrvivaz) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala:271]: deleted my previous comment... it is ambiguous b/c for full table scans, strategy.values will be None. So it means that either: it was not against a temporal index, or: it was against a temporal index, but was a full table scan. Since that ambiguity can't be easily resolved without refactoring our query plans, we probably would need to also configure block full table scans to catch that case.

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala [hunk 5]
```diff
@@ -160,14 +164,135 @@ object QueryInterceptor extends LazyLogging {
...
+          None       // Returning None means that the query continues.  Is this correct?
+        case Some(intervals) =>
+          if (!intervals.forall(_.isBoundedBothSides) || intervals.isEmpty) {
+            Some(new IllegalArgumentException("At least one part of the query is temporally unbounded.  Add bounded temporal constraints."))
+          } else {
+            val queryDuration = duration(intervals.values)
+            guardLimits.find( _.sizeLimit > spatialExtent).flatMap {
```
Comment (elahrvivaz) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala:278]: should this be `>=`? also, seems like there might not be a matching limit, in that case the query would go through, but seems like it shouldn't? since you would have configured a limit at a smaller size. (although also if the config is wrong the guards would be empty...)

Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala:278]: i was thinking `>=` would be needed to match full world scans

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/z2/package.scala (hunk 1 above)
Comment (jnh5y) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/z2/package.scala]: Good catch.  Just pushed a fix for this.

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala (hunk 2 above)
Comment (jnh5y) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala:207]: I like the pattern, and was thinking about the first time through; just didn't bother to figure out the syntax:).  I've added a little size check to handle the one limit case.  Maybe that case is degenerate and should throw some kind of error...  Thoughts?

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala (hunk 4 above)
Comment (jnh5y) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala:271]: Alright, this is complex enough and I'm forgetful enough that I'm writing a comment to note this information..

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala (hunk 2 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala:207]: i guess you were kind of assuming that there would be a config for the upper limit whole world bounding box, but it doesn't seem required... that probably breaks the logic in the query guard, as it would return None and let the query through (where you probably would expect it to fail?). so we should probably enforce something in the initialization or modify the logic in the guard.

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala (hunk 5 above)
Comment (jnh5y) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala:278]: You are correct.  I've added unit test cases to help with the corner cases.
The size of whole world guards is Max.Int:).

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/guard/TemporalQueryGuard.scala [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        s"Query exceeds maximum allowed filter duration of $max: ${filterToString(strategy.filter.filter)}")
+    }
+  }
+
+  override def close(): Unit = {}
+
+}
+
+object TemporalQueryGuard {
+  val Config = "geomesa.filter.max.duration"
```
Comment (jnh5y) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/guard/TemporalQueryGuard.scala:49]: @elahrvivaz after the refactoring, should this config path be different?

Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/guard/TemporalQueryGuard.scala:49]: sure, geomesa.guards.something.something?

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/guard/GraduatedQueryGuard.scala [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+  case class SizeAndDuration(sizeLimit: Int, durationLimit: Duration)
+
+  def loadConfiguration(sft: SimpleFeatureType): Seq[SizeAndDuration]  = {
+    val conf = ConfigFactory.load()
+    try {
+      val guardConfig = conf.getConfigList(s"${ConfigPath}.${sft.getTypeName}")
+      buildLimits(guardConfig)
+    } catch {
+      case e: ConfigException =>
+        logger.error(s"Configuration for GraduatedQueryGuard not available or correct for type ${sft.getTypeName}.  " +
```
Comment (elahrvivaz) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/guard/GraduatedQueryGuard.scala:104]: this is good, but FYI if the init method throws an exception, the framework will catch it and the guard won't even be invoked, so you don't need to ensure it still works if there's an error.

File: docs/user/datastores/query_interceptor.rst [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+      def rewrite(query: Query): Unit
+
+      /**
+       * Hook to allow interception of a query after extracting the query values
+       *
+       * @param strategy query strategy
+       * @return an exception if the query should be stopped
+       */
+      def guard(strategy: QueryStrategy): Option[IllegalArgumentException] = None
+    }
```
Comment (elahrvivaz) [inline docs/user/datastores/query_interceptor.rst:46]: i wonder if we should make this a java interface... and if that would break existing impls or not.

File: docs/user/datastores/query_interceptor.rst [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+Graduated Query Guard
++++++++++++++++++++++
+
+The graduated query guard applies different duration limits based on the spatial extent of the query.
+As a query becomes larger in space, it can be limited to shorter and shorter time ranges.
+A series of rules limit the duration for queries which are at most a given size in square degrees.
+
+To enable the guard, add ``org.locationtech.geomesa.index.planning.guard.GraduatedQueryGuard``
+to ``geomesa.query.interceptors`` as indicated above.  Configuration is managed via TypeSafe config.
```
Comment (elahrvivaz) [inline docs/user/datastores/query_interceptor.rst:91]: should probably clarify here that standard config loading is used and the config is expected under the key `geomesa.guard.graduated`

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/guard/GraduatedQueryGuard.scala [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+        if (first.sizeLimit == second.sizeLimit) {
+          throw new IllegalArgumentException(s"Graduated query guard configuration " +
+            s"has repeated size: ${first.sizeLimit}")
+        } else if (first.durationLimit.compareTo(second.durationLimit) <= 0) {
+          throw new IllegalArgumentException(s"Graduated query guard configuration " +
+            s"has durations out of order: ${first.durationLimit} is less than ${second.durationLimit}")
+        }
+      }
+    }
+    if (candidate.last.sizeLimit != Int.MaxValue) {
```
Comment (elahrvivaz) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/guard/GraduatedQueryGuard.scala:116]: should probably make this .lastOption.exists or add a .isEmpty check for the degenerate case of an empty Seq

File: geomesa-tools/conf/application.conf [hunk 11]
```diff
@@ -9,6 +9,31 @@ geomesa {
...
+    "example-csv-graduated-guard" = {
+      attributes = [
+        { name = "fid",      type = "Integer",         index = false }
+        { name = "name",     type = "String",          index = true  }
+        { name = "age",      type = "Integer",         index = false }
+        { name = "lastseen", type = "Date",            index = false, default = true }
+        { name = "geom",     type = "Point",           index = true, srid = 4326, default = true }
+      ]
+      user-data {
+        geomesa.query.interceptors = "org.locationtech.geomesa.index.planning.guard.GraduatedQueryGuard"
```
Comment (elahrvivaz) [inline geomesa-tools/conf/application.conf:34]: this is going to fail if you don't have the config file right? is that going to be confusing?

Comment (jnh5y) [reply geomesa-tools/conf/application.conf:34]: Yeah, I don't know what to do about it though...  I mean, having a user-data key-value would get around that....

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/guard/GraduatedQueryGuard.scala (hunk 10 above)
Comment (jnh5y) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/guard/GraduatedQueryGuard.scala:116]: Good call.  Adding an .isEmpty right ahead of this.

File: docs/user/datastores/query_interceptor.rst (hunk 9 above)
Comment (jnh5y) [reply docs/user/datastores/query_interceptor.rst:91]: Adding some docs for that.

File: docs/user/datastores/query_interceptor.rst (hunk 8 above)
Comment (jnh5y) [reply docs/user/datastores/query_interceptor.rst:46]: Hmm... do we have folks implementing the GeoMesa Java APIs?  Also, it should still be possible to do so...

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/guard/GraduatedQueryGuard.scala [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+        if (first.sizeLimit == second.sizeLimit) {
+          throw new IllegalArgumentException(s"Graduated query guard configuration " +
+            s"has repeated size: ${first.sizeLimit}")
+        } else if (first.durationLimit.compareTo(second.durationLimit) <= 0) {
+          throw new IllegalArgumentException(s"Graduated query guard configuration " +
+            s"has durations out of order: ${first.durationLimit} is less than ${second.durationLimit}")
+        }
+      }
+    }
+    candidate
```
[the remark this answers is not in the document]
Comment (jnh5y) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/guard/GraduatedQueryGuard.scala:122]: Added this.

File: geomesa-tools/conf/application.conf (hunk 11 above)
Comment (elahrvivaz) [reply geomesa-tools/conf/application.conf:34]: actually... this is a typesafe config file. you could add it in here.

Comment (jnh5y) [reply geomesa-tools/conf/application.conf:34]: I thought about that.  It seemed weird to toss this file into GeoServer.  But yeah, lemme do that.

Comment (elahrvivaz) [reply geomesa-tools/conf/application.conf:34]: true... but at least it will work with the CLI tools...

**Review by elahrvivaz — APPROVED** [review]
